llvm/labs/lab2/alexseev_danila/LoopPlugin.cpp

```diff
@@ -0,0 +1,72 @@
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/Passes/PassBuilder.h"
+#include "llvm/Passes/PassPlugin.h"
+
+struct LoopPlugin : public llvm::PassInfoMixin<LoopPlugin> {
+  llvm::PreservedAnalyses run(llvm::Function &F,
+                              llvm::FunctionAnalysisManager &FAM) {
+    llvm::LLVMContext &Context = F.getContext();
+    llvm::Module *ParentModule = F.getParent();
+
+    llvm::FunctionType *funcType =
+        llvm::FunctionType::get(llvm::Type::getVoidTy(Context), false);
+
+    llvm::LoopAnalysis::Result &LI = FAM.getResult<llvm::LoopAnalysis>(F);
+    for (auto *Loop : LI) {
+      llvm::IRBuilder<> Builder(Loop->getHeader()->getContext());
+      llvm::BasicBlock *Preheader = Loop->getLoopPreheader();
+      llvm::BasicBlock *ExitBlock = Loop->getExitBlock();
+
+      bool loopStartCalled = false;
+      for (auto &inst : *Preheader) {
+        if (auto *instCall = llvm::dyn_cast<llvm::CallInst>(&inst)) {
+          if (auto *CalledFunction = instCall->getCalledFunction()) {
+            if (CalledFunction->getName() == "loop_start") {
+              loopStartCalled = true;
+              break;
+            }
+          }
+        }
+      }

```

Comment:
Could you please make a separate helper function for that to reduce code duplication.